upgrades: upgrades don't seem to run when `developmentBranch` is set to `false`

When manually modifying `developmentBranch` (as in #99387) a couple of tests[1] fail. 

For example:

```
--- FAIL: TestServerStartupGuardrails (0.23s)
    test_server_shim.go:353: migration-manager-find-jobs: system-jobs-scan: relation "system.job_info" does not exist
```

This suggests that we are not properly running the SQL migrations in that case.

[1] Tests that fail:

- [x] TestAlreadyRunningJobsAreHandledProperly, TestPauseMigration - fixed by https://github.com/cockroachdb/cockroach/pull/100867
- [x] TestServerStartupGuardrails - fixed by https://github.com/cockroachdb/cockroach/pull/100729

Jira issue: CRDB-26555

3 tests that fail in https://github.com/cockroachdb/cockroach/pull/99387 are:

(1) `TestAlreadyRunningJobsAreHandledProperly`. in `pkg/upgrade/upgrademanager/upgrademanager_test`
```
=== RUN   TestAlreadyRunningJobsAreHandledProperly
    manager_external_test.go:120:
          Error Trace:  pkg/upgrade/upgrademanager/upgrademanager_test_test/pkg/upgrade/upgrademanager/manager_external_test.go:120
          Error:        Received unexpected error:
                        sql: no rows in result set
```
(2) `TestPauseMigration`. in `pkg/upgrade/upgrademanager/upgrademanager_test`
```
=== RUN   TestPauseMigration
    manager_external_test.go:464: error scanning '&{<nil> 0xc003f80200}': sql: no rows in result set
```
(3) `TestServerStartupGuardrails`. in `github.com/cockroachdb/cockroach/pkg/ccl/serverccl`
```
I230405 05:30:21.943950 189123 upgrade/upgrademanager/manager.go:168  [T1,n1] 1170  running permanent upgrades up to version: 1000022.2
I230405 05:30:21.944731 189876 sql/temporary_schema.go:505  [T1,n1] 1171  running temporary object cleanup background job
I230405 05:30:21.945053 189877 sql/sqlstats/persistedsqlstats/provider.go:129  [T1,n1] 1172  starting sql-stats-worker with initial delay: 10m15.992188656s
I230405 05:30:21.946647 189895 kv/kvclient/rangefeed/rangefeedcache/watcher.go:335  [T1,n1] 1173  system-config-cache: established range feed cache
I230405 05:30:21.947169 189864 kv/kvserver/replica_rangefeed.go:701  [T1,n1,rangefeed=table-stats-cache,mux_n=1,gen=2,r21/1:/Table/2{0-1},s1] 1174  RangeFeed closed timestamp is empty
I230405 05:30:21.950611 189876 sql/temporary_schema.go:555  [T1,n1] 1175  found 0 temporary schemas
I230405 05:30:21.950720 189876 sql/temporary_schema.go:558  [T1,n1] 1176  early exiting temporary schema cleaner as no temporary schemas were found
I230405 05:30:21.950747 189876 sql/temporary_schema.go:559  [T1,n1] 1177  completed temporary object cleanup job
I230405 05:30:21.950765 189876 sql/temporary_schema.go:640  [T1,n1] 1178  temporary object cleaner next scheduled to run at 2023-04-05 06:00:21.944702855 +0000 UTC m=+1944.305053323
E230405 05:30:21.994611 189894 spanconfig/spanconfigmanager/manager.go:141  [T1,n1] 1179  error starting auto span config reconciliation job: get-jobs: system-jobs-scan: relation "system.job_info" does not exist
I230405 05:30:22.008682 189123 upgrade/upgrademanager/manager.go:229  [T1,n1] 1180  the last permanent upgrade (v22.2-98) does not appear to have completed; attempting to run all upgrades
I230405 05:30:22.009905 189123 upgrade/upgrademanager/manager.go:270  [T1,n1] 1181  running permanent upgrade for version 0.0-2
    test_server_shim.go:331: migration-manager-find-jobs: system-jobs-scan: relation "system.job_info" does not exist
        (1) attached stack trace
          -- stack trace:
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal.func1.1

```

--- a/pkg/upgrade/upgrademanager/manager_external_test.go
+++ b/pkg/upgrade/upgrademanager/manager_external_test.go
@@ -56,9 +56,13 @@ func TestAlreadyRunningJobsAreHandledProperly(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	// We're going to be migrating from startCV to endCV.
-	startCV := roachpb.Version{Major: 41}
-	endCV := roachpb.Version{Major: 42}
+	// clusterversion.V23_1StopWritingPayloadAndProgressToSystemJobs was chosen
+	// specifically so that all the migrations that introduce and backfill the new
+	// `system.job_info` have run by this point. In the future this startCV should
+	// be changed to V23_2Start and updated to the next Start key everytime the
+	// compatability window moves forward.
+	startCV := clusterversion.V23_1StopWritingPayloadAndProgressToSystemJobs
+	endCV := startCV + 1
 
 	ch := make(chan chan error)
 
@@ -66,10 +70,11 @@ func TestAlreadyRunningJobsAreHandledProperly(t *testing.T) {
 	tc := testcluster.StartTestCluster(t, 2, base.TestClusterArgs{
 		ReplicationMode: base.ReplicationManual,
 		ServerArgs: base.TestServerArgs{
-			Settings: cluster.MakeTestingClusterSettingsWithVersions(endCV, startCV, false),
 			Knobs: base.TestingKnobs{
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
 				Server: &server.TestingKnobs{
-					BinaryVersionOverride:          startCV,
+					BootstrapVersionKeyOverride:    clusterversion.BinaryMinSupportedVersionKey,
+					BinaryVersionOverride:          clusterversion.ByKey(startCV),
 					DisableAutomaticVersionUpgrade: make(chan struct{}),
 				},
 				DistSQL: &execinfra.TestingKnobs{
@@ -77,11 +82,8 @@ func TestAlreadyRunningJobsAreHandledProperly(t *testing.T) {
 					ProcessorNoTracingSpan: true,
 				},
 				UpgradeManager: &upgradebase.TestingKnobs{
-					ListBetweenOverride: func(from, to roachpb.Version) []roachpb.Version {
-						return []roachpb.Version{to}
-					},
 					RegistryOverride: func(v roachpb.Version) (upgradebase.Upgrade, bool) {
-						if v != endCV {
+						if v != clusterversion.ByKey(endCV) {
 							return nil, false
 						}
 						return upgrade.NewTenantUpgrade("test", v, upgrade.NoPrecondition, func(
@@ -107,12 +109,25 @@ func TestAlreadyRunningJobsAreHandledProperly(t *testing.T) {
 	})
 	defer tc.Stopper().Stop(ctx)
 
+	// At this point the test cluster has run all the migrations until startCV.
+
 	upgrade1Err := make(chan error, 1)
 	go func() {
 		_, err := tc.ServerConn(0).ExecContext(ctx, `SET CLUSTER SETTING version = $1`, endCV.String())
 		upgrade1Err <- err
 	}()
 	unblock := <-ch
+
+	var firstID jobspb.JobID
+	var firstPayload, firstProgress []byte
+	require.NoError(t, tc.ServerConn(0).QueryRow(`
+   SELECT id, payload, progress FROM crdb_internal.system_jobs WHERE (
+                    crdb_internal.pb_to_json(
+                        'cockroach.sql.jobs.jobspb.Payload',
+                        payload
+                    )->'migration'
+                   ) IS NOT NULL AND status = 'running'
+`).Scan(&firstID, &firstPayload, &firstProgress))
 
 	// Inject a second job for the same upgrade and ensure that that causes
 	// an error. This is pretty gnarly.
@@ -125,21 +140,25 @@ func TestAlreadyRunningJobsAreHandledProperly(t *testing.T) {
                     unique_rowid(),
                     status,
                     created,
-                    payload,
-                    progress,
+                    NULL,
+                    NULL,
                     created_by_type,
                     created_by_id,
                     claim_session_id,
                     claim_instance_id
-              FROM system.jobs
-             WHERE (
-                    crdb_internal.pb_to_json(
-                        'cockroach.sql.jobs.jobspb.Payload',
-                        payload
-                    )->'migration'
-                   ) IS NOT NULL
+              FROM crdb_internal.system_jobs
+             WHERE id = $1
           )
-RETURNING id;`).Scan(&secondID))
+RETURNING id;`, firstID).Scan(&secondID))
+	// Insert the job payload and progress into the `system.job_info` table.
+	err := tc.Server(0).InternalDB().(isql.DB).Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		infoStorage := jobs.InfoStorageForJob(txn, secondID)
+		if err := infoStorage.WriteLegacyPayload(ctx, firstPayload); err != nil {
+			return err
+		}
+		return infoStorage.WriteLegacyProgress(ctx, firstProgress)
+	})
+	require.NoError(t, err)
 
 	// Make sure that the second job gets run in a timely manner.
 	runErr := make(chan error)
@@ -150,7 +169,7 @@ RETURNING id;`).Scan(&secondID))
 	fakeJobBlockChan := <-ch
 
 	// Ensure that we see the assertion error.
-	_, err := tc.Conns[0].ExecContext(ctx, `SET CLUSTER SETTING version = $1`, endCV.String())
+	_, err = tc.Conns[0].ExecContext(ctx, `SET CLUSTER SETTING version = $1`, endCV.String())
 	require.Regexp(t, "found multiple non-terminal jobs for version", err)
 
 	// Let the fake, erroneous job finish with an error.
@@ -392,9 +411,13 @@ func TestPauseMigration(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	// We're going to be migrating from startCV to endCV.
-	startCV := roachpb.Version{Major: 41}
-	endCV := roachpb.Version{Major: 42}
+	// clusterversion.V23_1StopWritingPayloadAndProgressToSystemJobs was chosen
+	// specifically so that all the migrations that introduce and backfill the new
+	// `system.job_info` have run by this point. In the future this startCV should
+	// be changed to V23_2Start and updated to the next Start key everytime the
+	// compatability window moves forward.
+	startCV := clusterversion.V23_1StopWritingPayloadAndProgressToSystemJobs
+	endCV := startCV + 1
 
 	type migrationEvent struct {
 		unblock  chan<- error
@@ -405,19 +428,16 @@ func TestPauseMigration(t *testing.T) {
 	tc := testcluster.StartTestCluster(t, 2, base.TestClusterArgs{
 		ReplicationMode: base.ReplicationManual,
 		ServerArgs: base.TestServerArgs{
-			Settings: cluster.MakeTestingClusterSettingsWithVersions(endCV, startCV, false),
 			Knobs: base.TestingKnobs{
 				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
 				Server: &server.TestingKnobs{
-					BinaryVersionOverride:          startCV,
+					BinaryVersionOverride:          clusterversion.ByKey(startCV),
+					BootstrapVersionKeyOverride:    clusterversion.BinaryMinSupportedVersionKey,
 					DisableAutomaticVersionUpgrade: make(chan struct{}),
 				},
 				UpgradeManager: &upgradebase.TestingKnobs{
-					ListBetweenOverride: func(from, to roachpb.Version) []roachpb.Version {
-						return []roachpb.Version{to}
-					},
 					RegistryOverride: func(cv roachpb.Version) (upgradebase.Upgrade, bool) {
-						if cv != endCV {
+						if cv != clusterversion.ByKey(endCV) {
 							return nil, false
 						}
 						return upgrade.NewTenantUpgrade("test", cv, upgrade.NoPrecondition, func(
@@ -454,13 +474,13 @@ func TestPauseMigration(t *testing.T) {
 	var id int64
 	tdb.QueryRow(t, `
 SELECT id
-  FROM system.jobs
+  FROM crdb_internal.system_jobs
  WHERE (
         crdb_internal.pb_to_json(
             'cockroach.sql.jobs.jobspb.Payload',
             payload
         )->'migration'
-       ) IS NOT NULL;`).
+       ) IS NOT NULL AND status = 'running';`).
 		Scan(&id)
 	tdb.Exec(t, "PAUSE JOB $1", id)
 

